File: app/src/main/java/megacorp/animals/MainActivity.java
package megacorp.animals;

import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.media.SoundPool;
import android.media.SoundPool.OnLoadCompleteListener;

import java.util.ArrayList;

public class MainActivity extends AppCompatActivity implements AnimalAdapter.OnAnimalClickListener {

    boolean loaded;
    SoundPool soundPool;
    ArrayList<Animal> animals = new ArrayList<>();

    // Todo convert all classes to kotlin

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        buildAnimalData();
        soundPool = buildSoundPool();

        AnimalAdapter animalAdapter = new AnimalAdapter();
        animalAdapter.setAnimals(animals);
        animalAdapter.setOnAnimalClickListener(this);

        RecyclerView recyclerView = (RecyclerView) findViewById(R.id.recyclerView);
        recyclerView.setLayoutManager(new GridLayoutManager(this, 2));
        recyclerView.setAdapter(animalAdapter);

    }

    public SoundPool buildSoundPool() {
        SoundPool.Builder builder = new SoundPool.Builder();
        builder.setMaxStreams(16);
        soundPool = builder.build();
        for (Animal animal : animals) {
            soundPool.load(this, animal.getSoundId(), 1);
        }
        soundPool.setOnLoadCompleteListener(new OnLoadCompleteListener() {
            public void onLoadComplete(SoundPool soundPool, int sampleId, int status) {
                loaded = true;
            }
        });
        return soundPool;
    }

    void buildAnimalData() {
        animals.add(new Animal(R.raw.cat, R.drawable.cat));
        animals.add(new Animal(R.raw.chicken2, R.drawable.chicken));
        animals.add(new Animal(R.raw.cow, R.drawable.cow));
        animals.add(new Animal(R.raw.dog1, R.drawable.dog));
        animals.add(new Animal(R.raw.duck, R.drawable.duck));
        animals.add(new Animal(R.raw.horse, R.drawable.horse));
        animals.add(new Animal(R.raw.pig, R.drawable.pig));
        animals.add(new Animal(R.raw.sheep, R.drawable.sheep));
        animals.add(new Animal(R.raw.turkey2, R.drawable.turkey));
        animals.add(new Animal(R.raw.mouse2, R.drawable.mouse));
    }

    @Override
    public void onAnimalClick(Animal item, int position) {
        if (loaded) {
            soundPool.play(position + 1, 1.0f, 1.0f, 1, 0, 1.0f);
        }
    }

}

File: app/src/main/java/megacorp/animals/MainActivity.kt
package megacorp.animals

import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.GridLayoutManager
import android.media.SoundPool
import androidx.recyclerview.widget.RecyclerView

import java.util.ArrayList

class MainActivity : AppCompatActivity(), AnimalAdapter.OnAnimalClickListener {

    private var loaded: Boolean = false
    private lateinit var soundPool: SoundPool
    private var animals = ArrayList<Animal>()
    private lateinit var soundPoolMap: HashMap<Int, Int>

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        buildAnimals()
        buildAndPopulateSoundPool()

        val animalAdapter = AnimalAdapter()
        animalAdapter.setAnimals(animals)
        animalAdapter.setOnAnimalClickListener(this)

        val recyclerView = findViewById<RecyclerView>(R.id.recyclerView)

        recyclerView.layoutManager =
            GridLayoutManager(
                this,
                this.resources.getInteger(R.integer.span_count)
            )
        recyclerView.adapter = animalAdapter
    }

    private fun buildAndPopulateSoundPool() {
        soundPool = SoundPool.Builder().setMaxStreams(16).build()
        soundPool.setOnLoadCompleteListener { _, _, _ -> loaded = true }
        soundPoolMap = HashMap()
        animals.forEach {
            soundPoolMap[it.soundId] = soundPool.load(this, it.soundId, 1)
        }
    }

    private fun buildAnimals() {
        animals.add(Animal(R.raw.cat, R.drawable.cat, "Cat"))
        animals.add(Animal(R.raw.chicken2, R.drawable.chicken, "Chicken"))
        animals.add(Animal(R.raw.cow, R.drawable.cow, "Cow"))
        animals.add(Animal(R.raw.dog1, R.drawable.dog, "Dog"))
        animals.add(Animal(R.raw.duck, R.drawable.duck, "Duck"))
        animals.add(Animal(R.raw.horse, R.drawable.horse, "Horse"))
        animals.add(Animal(R.raw.pig, R.drawable.pig, "Pig"))
        animals.add(Animal(R.raw.sheep, R.drawable.sheep, "Sheep"))
        animals.add(Animal(R.raw.turkey2, R.drawable.turkey, "Turkey"))
        animals.add(Animal(R.raw.mouse2, R.drawable.mouse, "Mouse"))
    }

    override fun onAnimalClick(item: Animal) {
        if (loaded && soundPoolMap.containsKey(item.soundId)) {
            val soundId: Int = soundPoolMap[item.soundId] ?: return
            soundPool.play(soundId, 1.0f, 1.0f, 1, 0, 1.0f)
        }
    }

}

data class Animal(val soundId: Int, val imageId: Int, val animalName: String)

